### ISSUE TITLE
Facade classification

### ISSUE THREAD

**Issue by vvoovv** [prochitecture/blosm#8]
I started the [FacadeClassification](https://github.com/prochitecture/blosm/blob/dev/action/facade_classification.py) class.

**Comment by vvoovv** [comment]
Several iteration passes `for vector in building.polygon.getVectors()` will be probably needed.

**Comment by vvoovv** [comment]
Use `vector.next` or `vector.prev` to access a neighbor vector.

**Comment by vvoovv** [comment]
It looks like a bug.

A number of shared edges (marked with the blue color) are classified as the front facades for _husum_altstadt.osm_.

![image](https://user-images.githubusercontent.com/613295/116881489-fcd01a80-ac2b-11eb-99e7-39332efd704e.png)

**Comment by polarkernel** [comment]
> A number of shared edges (marked with the blue color) are classified as the front facades for husum_altstadt.osm.

A first bug fix for the larger edge at the right of the image (see image below). 

<img src="https://user-images.githubusercontent.com/48180714/117273654-a0a00d00-ae5c-11eb-9c99-1217dfaf6428.png" width="400">

Although it should be a front facade, it has a visibility of zero (see the numbers in the image). Like this, the building had no visible edges at all (all others are shared edges) and was classified using the rule for 'invisible' buildings: Find the longest edge and mark it as front edge. And this was a bug, because shared edges have to be excluded from this search. It is fixed now, but not yet committed.

However, it is only a symptom for another issue: Why is this edge not visible? I found the reason by a brute force debugging. In the image below, every way-segment, associated in _visInfo_ to an edge that has a non-zero visibility, is marked by an arrow (magenta) pointing from the middle of the segment to its edge:

<img src="https://user-images.githubusercontent.com/48180714/117275008-e5787380-ae5d-11eb-8367-3cc2aecb819b.png" width="800">

The dotted green line shows the border of the search range, currently without any margin to the way-segment. It narrowly misses this edge, while the next segment is already tilted too much, so that its search-range again misses this edge. It is an example for possible issues when not using a margin. I already pointed once to this problem when I explained the reason for the 10m margin in [Discussions](https://github.com/prochitecture/blosm/discussions/1#discussioncomment-464703). 

It is also interesting how the order of processing the way-segments influences these associations. If you follow the front edges from the edge I discussed above to the left, there are not all edges associated to the long way-segment. Two of them seem to be processed by the short segment at the left, before the long segment.

I continue searching for solutions (and for other bug fixes).

**Comment by vvoovv** [comment]
> In the image below, every way-segment, associated in _visInfo_ to an edge that has a non-zero visibility, is marked by an arrow (magenta) pointing from the middle of the segment to its edge:

Very useful visualization. Can we have it in the module _mpl.renderer.facade_classification_?

**Comment by vvoovv** [comment]
> It is also interesting how the order of processing the way-segments influences these associations. If you follow the front edges from the edge I discussed above to the left, there are not all edges associated to the long way-segment. Two of them seem to be processed by the short segment at the left, before the long segment.

How about using the angle between the building edge and a way segment in the method `VisibilityInfo.__gt__`?

**Comment by vvoovv** [comment]
> How about using the angle between the building edge and a way segment in the method `VisibilityInfo.__gt__`?

A way-segment with the angle to the building edge that is closer to zero would win.

**Comment by vvoovv** [comment]
> It is an example for possible issues when not using a margin. I already pointed once to this problem when I explained the reason for the 10m margin in [Discussions](https://github.com/prochitecture/blosm/discussions/1#discussioncomment-464703).

A possible solution can be using the margin, detect T-crossings and skip the buildings to the left from the base of the 'T'. Detection of T-crossings will be needed anyway for the generation of realistic streets.

<img src="https://user-images.githubusercontent.com/48180714/110679339-a9b38b80-81d7-11eb-85d0-26bd2b3bb21e.png" width="300">

**Comment by polarkernel** [comment]
> A first bug fix for the larger edge at the right of the image (see image below).

Bug fixed and committed.

> Very useful visualization. Can we have it in the module mpl.renderer.facade_classification?

Committed. Use the command line parameter _--showAssoc_, available for _BuildingVisibilityRender_ and _BuildingClassificationRender_.

**Comment by polarkernel** [comment]
> How about using the angle between the building edge and a way segment in the method VisibilityInfo.__gt__?
A way-segment with the angle to the building edge that is closer to zero would win.

The association between the way-segement and the edge is currently not used (beside the check for the way-category). So this feature would not change anything. However, I will test a solution, there's nothing wrong to get more order.

> A possible solution can be using the margin, detect T-crossings and skip the buildings to the left from the base of the 'T'. Detection of T-crossings will be needed anyway for the generation of realistic streets.

T-crossings are not the only issue, also dead-ends can produce such issues, as once "solved" [here](https://github.com/prochitecture/blosm/discussions/1#discussioncomment-655258).

What about making the margins adaptive to the angle between the way segments, using some indicator for this angle, something like the absolute value of a dot product. For large angles, like in the above issue, the margin would be wider, smaller for small angles and zero at the ends of a way. Do you think it would be very complex to provide values for these angles (not necessarily linear) for both sides of a way-segment?

**Comment by vvoovv** [comment]
> What about making the margins adaptive to the angle between the way segments, using some indicator for this angle, something like the absolute value of a dot product. For large angles, like in the above issue, the margin would be wider, smaller for small angles and zero at the ends of a way. Do you think it would be very complex to provide values for these angles (not necessarily linear) for both sides of a way-segment?

What if there are 3 or more way-segments sharing a node?

Did you mean the normalized absolute value of the dot product, i.e. the absolute value of the cosine between the related way-segments?

**Comment by polarkernel** [comment]
> What if there are 3 or more way-segments sharing a node?

There can't be a perfect solution in my opinion. I would just take the linear way as defined by OSM and set the angle to zero at its ends.

> Did you mean the normalized absolute value of the dot product, i.e. the absolute value of the cosine between the related way-segments?

I thought on a even simpler solution. In the _bpypolyskel_ project I used a _pseudoangle_, defined as follows for a vector _d_ (the vector from the start to the end of a way-segment):

```Python
# Adapted from https://stackoverflow.com/questions/16542042/fastest-way-to-sort-vectors-by-angle-without-actually-computing-that-angle
# Input:  d: difference vector.
# Output: a number from the range [0 .. 4] which is monotonic
#         in the angle this vector makes against the x axis.
def pseudoangle(d):
    p = d[0]/(abs(d[0])+abs(d[1])) # -1 .. 1 increasing with x
    if d[1] < 0: 
        return 3 + p  #  2 .. 4 increasing with x
    else:
        return 1 - p  #  0 .. 2 decreasing with x
```

Something similar could be used to compute the angle of a way-segment relative to the x-axis. The differences of this angle to the values of the _prev_ and _next_ of the way-segment would already be a good approximations, in my opinion.

Another very simple solution would use a zero margin at the ends of a way and a constant margin for all segments in between. Maybe this could be a starting point for experiments. At least it would solve the dead-end and T-junction problems.

**Comment by vvoovv** [comment]
> Another very simple solution would use a zero margin at the ends of a way and a constant margin for all segments in between. Maybe this could be a starting point for experiments. At least it would solve the dead-end and T-junction problems.

I'd like to note that it's a more complex problem. Two ways may share a single node. A way may turn left or right at a T-junction. I'll add a _Junction_ class. 

![image](https://user-images.githubusercontent.com/613295/117453490-b04b4e80-af4d-11eb-8ada-448fdf2207a7.png)

![image](https://user-images.githubusercontent.com/613295/117451665-7711df00-af4b-11eb-8939-82f2c55d4fb8.png)

**Comment by vvoovv** [comment]
Also I'll add the code to treat the case of a node shared by two different ways.

**Comment by polarkernel** [comment]
> I'd like to note that it's a more complex problem. Two ways may share a single node. A way may turn left or right at a T-junction. I'll add a Junction class.

I see, my thinking was once again too easy.

The new visualization discovered more problems already on the determination of visibility. Way-segments in tunnels behave different since we introduced shared edges, as shown below. I try to fix this issue.

<img src="https://user-images.githubusercontent.com/48180714/117462694-3320d900-af4f-11eb-9c96-3b2fa2ff0224.png" width="400">

**Comment by vvoovv** [comment]
> The new visualization discovered more problems already on the determination of visibility. Way-segments in tunnels behave different since we introduced shared edges, as shown below. I try to fix this issue.

Where this case is located?

The class _way.Way_ have 2 boolean attributes: _tunnel_ and _bridge_. I guess the majority of tunnels and bridges can be safely skipped from the visibility calculation. But there could be exceptions. There are even bridges serving also as buildings, for example [Ponte Vecchio](https://en.wikipedia.org/wiki/Ponte_Vecchio).

Anyway I skipped bridges and tunnels from the visibility calculations right at [way.manager.WayManager.getFacadeVisibilityWays(..)](https://github.com/prochitecture/blosm/blob/9adbadd3020b2668895ef47f4abf39ecc72bdee5/way/manager.py#L37).

**Comment by polarkernel** [comment]
> Way-segments in tunnels behave different since we introduced shared edges, as shown below.
> Where this case is located?

It is in _husum_altstadt.osm_ immediately at the right of the place I discussed the other problem [above](https://github.com/prochitecture/blosm/issues/8#issuecomment-833409159). However, in the meantime I found that it is not a bug. The way that goes through the tunnel there really sees a very small part of this edge and produces a visibility of 0.01.

> Anyway I skipped bridges and tunnels from the visibility calculations right at way.manager.WayManager.getFacadeVisibilityWays(..).

Does that mean that we skip the requirement that [passages](https://github.com/prochitecture/blosm/discussions/1#discussioncomment-349547) should have front facades?

**Comment by vvoovv** [comment]
> It is in _husum_altstadt.osm_ immediately at the right of the place I discussed the other problem [above](https://github.com/prochitecture/blosm/issues/8#issuecomment-833409159). 

But there is a building in that place. Did you remove it for testing?
![image](https://user-images.githubusercontent.com/613295/117485384-f9f96080-af70-11eb-91f8-ca66daff0a2e.png)


> Does that mean that we skip the requirement that [passages](https://github.com/prochitecture/blosm/discussions/1#discussioncomment-349547) should have front facades?

No. Only the tunnels are skipped. The ways leading to the tunnels do take part in the visibility calculation.

**Comment by polarkernel** [comment]
> But there is a building in that place. Did you remove it for testing?

You are right. It seems that it's now me that has this magic behavior. In a recent run, the building is there again and the visibility has gone. But what could make different results using identical code? Very strange.

**Comment by polarkernel** [comment]
> You are right. It seems that it's now me that has this magic behavior. In a recent run, the building is there again and the visibility has gone. But what could make different results using identical code? Very strange.

The reason for the magic behavior is a person: Me! I divided _husum_altstadt.osm_ into a couple of smaller files to ease debugging. For the above test I used accidentally a file, where this building was deleted, while the _unsaved_ project file showed another one to me. Sorry for the confusion.

**Comment by polarkernel** [comment]
> A way-segment with the angle to the building edge that is closer to zero would win.

I tested a first version of weighting on the comparison of two way-segments using _\_\_gt\_\_()_ in _VisibilityInfo_, using four goals:

- Prefer way-segments that produce more visibility
- Prefer way-segments that are more parallel to the egde (your proposition)
- Prefer way-segments that are closer to the edge
- Prefer way-segments that have a more important way-category

Visibility can immediately used as a weight. For the angle condition I used _dx_ and _dy_ and an equation derived from the pseudoangles:

`w_angle = dx/(dx+dy)`

This produces a continuous value between 0 and 1, which is not linear, but useful for comparisons:

<img src="https://user-images.githubusercontent.com/48180714/117544863-a3009380-b023-11eb-96d9-8f4b69b11e32.png" width="400">

As weight for the distance I used the distance `d = abs(Y1+Y2)` provided in _VisibilityInfo_  and tried to create a weight that linearly decreases with _d_:

`w_dist = (100.-info.distance/2.)/100. if info.distance/2. < 100. else 0.`

The value 100 is the height range of the search rectangle. The distance is divided by 2 because it is a sum of two distances. Finally, the importance of the way-segment is simply weighted as a constant:

`w_way = 0.5 if WayLevel[info.waySegment.way.category] == 1 else 0.0`

The final weight is curently just the sum of all weights:

`weight = w_angle + w_dist + w_way + visibility`

Every weight could be multiplied by a constant to distribute weights differently, but this should be the goal of experiments. Here a first impression of a region around the inner harbour of Husum (left: visibility only, right: using weights):

<img src="https://user-images.githubusercontent.com/48180714/117545557-96316f00-b026-11eb-96e1-5e24246089f0.png" width="400"> <img src="https://user-images.githubusercontent.com/48180714/117545632-e27caf00-b026-11eb-94f3-1c59b5cbb0fc.png" width="400">

The weighting worked best, when only applied to new way-segments that produce equal or more visibility. This solution is committed and ready for tests.

**Comment by polarkernel** [comment]
I fixed a bug at the insertion of crossed edges. Additionally, links to facades at dead-ends of way-segments are now plotted in blue when the command line parameter _--showAssoc_ is applied. This new version is commited.

**Comment by vvoovv** [comment]
My concern is still the case of the corner facade in _manhattan_01.osm_. Why does the magenta arrows point to the vertices of the building edge instead of its center? Why are there 2 magenta to that building edge instead of one?
Finally I'd classify that building edge the the front one.

![image](https://user-images.githubusercontent.com/613295/118371171-fe96c800-b5b3-11eb-8b69-cd4d4cab8216.png)

![image](https://user-images.githubusercontent.com/613295/118371258-75cc5c00-b5b4-11eb-98ca-c263ba1013c5.png)

**Comment by polarkernel** [comment]
> My concern is still the case of the corner facade in manhattan_01.osm. Why does the magenta arrows point to the vertices of the building edge instead of its center? Why are there 2 magenta to that building edge instead of one?

The magenta arrows always point to the middle of a building edge. In this example, there are very small edges (~10cm) on both sides of the corner facade of this building, which gets visible when magnified:

<img src="https://user-images.githubusercontent.com/48180714/118450203-3a38ab80-b6f4-11eb-90af-48d1d694757e.png" width="400">

> Finally I'd classify that building edge the the front one.

Note that the ways on both sides of the edge in question are not perpendicular. The smallest angle, determined by _dx_, _dy_, is from the way-segment at the right of the edge and has a value of 52°, which is just above the current limit of 50°, defined as _VisibilityAngle_ in _facade_classification.py_. Increasing this limit to 55° for example classifies this edge as front facade. We may try to increase this limit and see, whether we observe side effects.

**Comment by vvoovv** [comment]
> The magenta arrows always point to the middle of a building edge. In this example, there are very small edges (~10cm) on both sides of the corner facade of this building, which gets visible when magnified

Sorry, I didn't notice that.

> Note that the ways on both sides of the edge in question are not perpendicular. The smallest angle, determined by _dx_, _dy_, is from the way-segment at the right of the edge and has a value of 52°, which is just above the current limit of 50°, defined as _VisibilityAngle_ in _facade_classification.py_. Increasing this limit to 55° for example classifies this edge as front facade. We may try to increase this limit and see, whether we observe side effects.

I'd like to try a different idea. I'll report about the result tomorrow.

**Comment by vvoovv** [comment]
I moved the angle check from _facade_visibility.py_:
```python
if VisibilityAngleFact*dx < dy: 
    _visInfo.value = 0.
```
to _facade_classification.py_:
```python
        for way_level in range(1,MaxWayLevel+1): # way-level corresponds to way-categories C1, C2, C3
            for vector in building.polygon.getVectors():
                edge = vector.edge
                visInfo = edge.visInfo
                if visInfo.value and \
                        VisibilityAngleFact*visInfo.dx > visInfo.dy and \
                        WayLevel[visInfo.waySegment.way.category] == way_level:
```

Is it ok if I merge it to the dev _branch_?

**Comment by polarkernel** [comment]
> Is it ok if I merge it to the dev branch?

I am not sure whether this will work, at least it may produce different results. Let me assume two way-segments in the loop of segments in _facade_visibility.py_, applied to the same edge:

first segment:  visibility 1.0, angle 80°
later segment: visibility 0.9, angle 40°

In the current version, the visibility for the first segment will be set to 0. because of the angle limit and the later segment wins and gets updated. In the new version, the first segment with visibility 1. gets updated, while the later segment looses, bcause of lower visibility. But the first version is the desired one, isn't it?

**Comment by vvoovv** [comment]
> I am not sure whether this will work, at least it may produce different results.

But we have the `VisibilityInfo.weight(..)` method to get the required result.

**Comment by polarkernel** [comment]
> But we have the VisibilityInfo.weight(..) method to get the required result.

OK, when we give priority of the angle over visibility in _VisibilityInfo.weight(..)_, it could work. Let's just try, we can always undo it.

**Comment by vvoovv** [comment]
Ok, I'll merge it to _dev_.

**Comment by vvoovv** [comment]
I moved all definitions to the modules _defs.facade_classification_ and _defs.way_

**Comment by vvoovv** [comment]
I set `VisibilityAngle = 50` again.

**Comment by vvoovv** [comment]
I'd expect that the building edge in the middle of the image below should be classified as the front one. The winning way segment would be the major street above the building edge rather than the service road.

![image](https://user-images.githubusercontent.com/613295/118975357-9ada1980-b97c-11eb-8229-abf3e5374eda.png)

![image](https://user-images.githubusercontent.com/613295/118976815-20aa9480-b97e-11eb-801e-202d8db6d90a.png)

**Comment by polarkernel** [comment]
> I'd expect that the building edge in the middle of the image below should be classified as the front one. The winning way segment would be the major street above the building edge rather than the service road.

This edge gets classified already in _facade_visibility.py_ as _deadend_, as it gets crossed by the axis of the way-segment of the service road within a distance of 10m. Because the building already got front facades by ways of more importance, its class gets reset to _unknown_ in lines [77-85](https://github.com/prochitecture/blosm/blob/43bda4bc0053747094a4f19dfbd6dbe1756401f9/action/facade_classification.py#L77-L85). Later, it becomes a side facade.

Crossing edges can't be processed in a normal way, because their angles and distances do not fit to these rules. At this point, it is also no more possible to find the winning way-segment as you propose. Should we maybe leave it as _deadend_ in this case?

**Comment by vvoovv** [comment]
> Should we maybe leave it as _deadend_ in this case?

I'd leave it as is for now.

**Comment by vvoovv** [comment]
The latest OSM file _sirius.osm_ produces an error. I don't understand how it can happen. How can an edge be classified and not have _waySegment_?

**Comment by polarkernel** [comment]
> The latest OSM file _sirius.osm_ produces an error. I don't understand how it can happen. How can an edge be classified and not have waySegment?

It seems the edge was classified as the longest edge of a building by _frontOfInvisibleBuilding()_. In the last commit I added a condition that requests the existance of the way-segments for the two edges.

**Comment by vvoovv** [comment]
> It seems the edge was classified as the longest edge of a building by _frontOfInvisibleBuilding()_.

In that case only one edge is classified as the front one but in the condition
```python
                        vector.prev.edge.cl == FacadeClass.front and \
                        vector.next.edge.cl == FacadeClass.front:
```
two facades must be the front ones.

**Comment by polarkernel** [comment]
> In that case only one edge is classified as the front one but in the condition
                        vector.prev.edge.cl == FacadeClass.front and \
                        vector.next.edge.cl == FacadeClass.front:
two facades must be the front ones.

You are right. I found the real reason for the bug. The indent of the block at line [85](https://github.com/prochitecture/blosm/blob/bbff6802bbc88df1a638c3b2891eba47ca05a55e/action/facade_classification.py#L86) was too large. In the first run of the loop for the way-categories, the winner segment of the edge with  the id 5064 had a lower category and was skipped. Then, in the block mentioned above, the segment with id 5062 was the longest and therefore selected as front. When the way-category loop did its second run with a less important way category, the segment 5064 was ordinarily set as front.

The block starting at line [85](https://github.com/prochitecture/blosm/blob/bbff6802bbc88df1a638c3b2891eba47ca05a55e/action/facade_classification.py#L86) required less indentation to be exectuted only once at the end of the function. This is fixed now and committed.

**Comment by vvoovv** [comment]
I successfully tested the following processing pipeline on the OSM file _feature_detection/tula_bay_windows_01.osm_:
Feature detection -> Feature removal -> Visibility Calculation ->Facade Classification

I encountered a couple of problems with the service roads. Note the marked footprint edge with the _id_ equal to 730.

![image](https://user-images.githubusercontent.com/613295/129228776-23e73c13-dc97-4d4e-b029-4d23c5ffcf64.png)

(1)
The winning way-segment is perpendicular to the the building edge. Shouldn't the way-segment parallel to the building edge be the winning one?

(2)
The marked footprint edge is classified as the front one with the winning way-segment representing a _service_ road. Actually it's should be classified as the back one. My proposal is to consider that a building edge with a _service_ winning way-segment has no winning edge at all if at there is at least one building edge that is
1. classified as the front one
2. has a winning way-segment other than _service_.

**Comment by vvoovv** [comment]
> I successfully tested the following processing pipeline on the OSM file _feature_detection/tula_bay_windows_01.osm_:
> Feature detection -> Feature removal -> Visibility Calculation ->Facade Classification

Here is how to reproduce the result:
```
python.exe -m script.command_line --buildings --highways --detectFeatures --simplifyPolygons --classification --setupScript setup\mpl_facade_visibility.py --osmFilepath ..\osm_extracts\feature_detection\tula_bay_windows_01.osm
```

**Comment by polarkernel** [comment]
> (1) The winning way-segment is perpendicular to the the building edge. Shouldn't the way-segment parallel to the building edge be the winning one?

This edge gets classified as _deadend_ already in _facade_visibility.py_. As _deadend_, it gets excluded from the standard classicfication process _classifyFrontFacades()_ in _facade_classification.py_ and, as it belongs to the group _CrossedFacades_, it gets classified as _front_ in  without further considerations (line [25](https://github.com/prochitecture/blosm/blob/bf405f563f54be679fcf2ab4541d8380ddb0a561/action/facade_classification.py#L25)). This explains why the perpendicular way-segment is associated to this edge.

> (2) The marked footprint edge is classified as the front one with the winning way-segment representing a service road. Actually it's should be classified as the back one.

Without the classification as _deadend_, this edge would not be classified as _front_ facade. I propose to remove the exclusion from _classifyFrontFacades()_ for _CrossedFacades_ edges.  These should then only get _front_ edges, when classified as such in _classifyFrontFacades()_, or when their way-segment level is at least as high as an already detected _front_ egde of the building, if any, or finally when no other _front_ edge was found for the building.

I try to implement and test this approach.

**Comment by polarkernel** [comment]
> I try to implement and test this approach.

Solution tested and committed. Seams to work.

**Comment by vvoovv** [comment]
Yes, now it outputs the expected result.

**Comment by vvoovv** [comment]
I'd like to share the initial results of integration of the facade classification processing pipeline with Blender.

The green color is used for front facades.

There are very narrow side facades in the screenshots marked with the yellow color.

The blue color is used for building parts. Building parts will somehow inherit the class of the footprint of the whole building.

The gray color is used for shared facades.

I was afraid how the Blender's KD-tree would behave. Likely it seemingly delivers the same results as scipy's KD-tree.

![image](https://user-images.githubusercontent.com/613295/150647771-39fb22c9-7a6a-4bd3-9acb-45af07b9263c.png)


![image](https://user-images.githubusercontent.com/613295/150647746-c20224a2-7081-4193-993a-053fe23f679a.png)

**Comment by polarkernel** [comment]
> I'd like to share the initial results of integration of the facade classification processing pipeline with Blender.

Looks fine. Interesting, all the theoretical considerations we made now become realistically visible.

> I was afraid how the Blender's KD-tree would behave.

You have nerves of steel if you have tried this for the first time only now! Glad to hear that it works.

**Comment by vvoovv** [comment]
I am currently writing the code to inherit the facade class for a building part from the one for the whole building footprint.

**Comment by vvoovv** [comment]
I've discovered an issue with passages. File _facade_visibility/manhattan_01.osm_. Vector ID is 42.

I think the facade should be classified as a front one.

![image](https://user-images.githubusercontent.com/613295/163257224-0433a72c-fc4f-472f-b9ba-80b31d92d79c.png)

![image](https://user-images.githubusercontent.com/613295/163257063-bb2248a5-3e0b-422a-9501-a3ef334a4e4e.png)

**Comment by polarkernel** [comment]
> I think the facade should be classified as a front one.

I am no more sure, in which repo did you observe this?

**Comment by vvoovv** [comment]
In the branch _dev_.

**Comment by polarkernel** [comment]
> I think the facade should be classified as a front one.

I am not sure how to solve that. The way that is perpendicular to the vector with id 42 should normally classify these vector's edge as _passage_ in the visibility computation. The _passage_ resolving algo in the facade classification should then change this class to _front_, because of the higher category of the Beaver Street. But without this _passage_ classification, the perpendicular way dominates, because it is next to the edge.

The classification to _passage_ requires an intersection of the perpendicular way with the facade vector (42). The vector (42)  originally consisted of two edges, that were combined to one by the straight-angle process. But now this combined edge is no more intersected by the way. In the following image, the original building is shown in yellow, the straightened edge is in red and the perpendicular way is the black line from bottom to the black dot. The black dot is also the vertex, where the two original edges met.

<img src="https://user-images.githubusercontent.com/48180714/163420202-b08cb793-d8da-4fac-ab26-11eb255e808d.png" width="300">

I could now elongate the ways a bit for the intersection test, but I don't know what side effects would be produced by that.

**Comment by vvoovv** [comment]
I don't understand why the vector (42) gets the _back_ class.

It should get the _front_ class even if the parallel street (Beaver street) isn't present at all.

**Comment by vvoovv** [comment]
The perpendicular edge is a _service_ road.

My proposal is to skip all service roads if there is at least one building edge that is
* got the _front_ class
* has a winning way-segment other than _service_

**Comment by polarkernel** [comment]
I made a mistake in debugging, the perpendicular edge **is** classified as _passage_ and therefore the service road was set as winning road in the visibility computation.

> I don't understand why the vector (42) gets the back class.

This is due to the rule we introduced in the second paragraph of [this comment](https://github.com/prochitecture/blosm/issues/8#issuecomment-898412230). The _passage_ edge was related to the perpendicular way, which has a lower level than ways of the edges of the building already classified as _front_.

> My proposal is to skip all service roads if there is at least one building edge that is
    got the front class
    has a winning way-segment other than service

I fixed this in the visibility computation. A way can't be classified as _passage_, if it already has a winning way-segment other than _service_. This fix is committed to the branch _dev_.

**Comment by vvoovv** [comment]
Thanks. Now the facades are classified as expected.
